### ISSUE TITLE
Networking is not working as expected

### ISSUE THREAD

**Issue by russell-islam** [cloud-hypervisor/libvirt#84]
After creating the guest VM sometimes I can not bring up the interface with dhclient , rarely works after changing the MAC. Also, I can not bring up the networking at all for the second guest on the same host.

**Comment by russell-islam** [comment]
The first guest network comes up sometimes using `dhclient ens4`. But it hangs after reboot.  Also for the 2nd guest it is not coming up at all. Looks like there is some root issues with the networking support we added before.

**Comment by russell-islam** [comment]
I did some tcpdump analysis, I can see host bridge sends some reply but anyhow dhclient is not able to process it. I am not sure if anything is missing in the libvirt code.

**Comment by gbalane** [comment]
Can you please attach the screen log here to understand. @vineethrp

**Comment by bryteise** [comment]
I'll start taking a look at this next week.

**Comment by gbalane** [comment]
Is the Libvirt CI that covers these multiple VM spun up with network interface? The tests are part of Kubernetes conformance tests.

**Comment by bryteise** [comment]
Yes but the current implementation has to run the tests in serial because they each use their own libvirtd daemon that isn't able to be run per thread.

**Comment by russell-islam** [comment]
My test: I brought up first vm worker1 just fine. Tried 2nd vm failed. Herer are the screen shots
# Bringing up first guest
![worker1](https://user-images.githubusercontent.com/13595257/115094516-b613be00-9ed2-11eb-93a0-f7bb836919ce.JPG)
# First VM cofnig
![worker2](https://user-images.githubusercontent.com/13595257/115094517-b6ac5480-9ed2-11eb-8a4c-f75a532e6ce0.JPG)
# Send guest IP configs and dhclient hangs
![master1](https://user-images.githubusercontent.com/13595257/115094519-b744eb00-9ed2-11eb-8549-611d6f47fe28.JPG)
# 2nd guest xml config
![master-xml](https://user-images.githubusercontent.com/13595257/115094520-b7dd8180-9ed2-11eb-823f-599db68cf657.JPG)
# Tcp dump of the host vnet1(2nd guest)
[tcpdump.txt](https://github.com/cloud-hypervisor/libvirt/files/6328287/tcpdump.txt)

**Comment by russell-islam** [comment]
# First VM got the IP address after successfully running dhclient on ens4 interface.
![workerip](https://user-images.githubusercontent.com/13595257/115094681-46520300-9ed3-11eb-88cc-fcc077945dd3.JPG)

**Comment by russell-islam** [comment]
# Host vnets after creating two VMs
![host-vnets](https://user-images.githubusercontent.com/13595257/115094719-684b8580-9ed3-11eb-9375-94d79705ecaa.JPG)

**Comment by russell-islam** [comment]
I think we have some missing functionalities there to work with dhcp. I got the following error while running.

`
 sudo virsh -c CH:///system net-update default add ip-dhcp-host "<host mac='a4:bb:6d:43:5d:9e' name='cbl-mariner-desktop-host' ip='192.168.122.45' />" --live --config
 
 
muislam@cbl-mariner-desktop-host [ /etc/libvirt/qemu/networks ]$ sudo virsh -c CH:///system net-update default add ip-dhcp-host "<host mac='a4:bb:6d:43:5d:9e' name='cbl-mariner-desktop-host' ip='192.168.122.45' />" --live --config
error: failed to get network 'default'
error: this function is not supported by the connection driver: virNetworkLookupByName

muislam@cbl-mariner-desktop-host [ /etc/libvirt/qemu/networks ]$
`

**Comment by rbradford** [comment]
For clarification: Are you expecting the virtio-net interfaces for each VM to be connected to the same bridge with a dhcp server running against that bridge with all the VMs getting dhcp leases from the same subnet?

**Comment by rbradford** [comment]
Manual use of bridging and dnmasq results in working guest DHCP for me:

```
sudo ip tuntap add tap0 mode tap
sudo ip tuntap add tap1 mode tap
sudo brctl addbr br0
sudo ip address add 192.168.100.1/24 dev br0
sudo ip link set br0 up
sudo brctl addif br0 tap0
sudo brctl addif br0 tap1
sudo dnsmasq -i br0 -a 192.168.100.1 -I lo -d -z -C /dev/null -F "192.168.100.2,192.168.100.10"
```

`
target/debug/cloud-hypervisor --kernel ~/src/linux/vmlinux --pmem file=~/workloads/focal1.raw --cpus boot=1 --memory size=512M --cmdline "root=/dev/pmem0p1 console=ttyS0"  --serial tty --console off --api-socket=/tmp/api1 --seccomp false --net tap=tap0`

`target/debug/cloud-hypervisor --kernel ~/src/linux/vmlinux --pmem file=~/workloads/focal2.raw --cpus boot=1 --memory size=512M --cmdline "root=/dev/pmem0p1 console=ttyS0"  --serial tty --console off --api-socket=/tmp/api1 --seccomp false --net tap=tap1`

So I think this must be libvirt related not CH related.

**Comment by liuw** [comment]
> I think we have some missing functionalities there to work with dhcp. I got the following error while running.
> 
> `
> sudo virsh -c CH:///system net-update default add ip-dhcp-host "" --live --config
> 
> muislam@cbl-mariner-desktop-host [ /etc/libvirt/qemu/networks ]$ sudo virsh -c CH:///system net-update default add ip-dhcp-host "" --live --config
> error: failed to get network 'default'
> error: this function is not supported by the connection driver: virNetworkLookupByName
> 
> muislam@cbl-mariner-desktop-host [ /etc/libvirt/qemu/networks ]$
> `

Please check if https://github.com/liuw/libvirt/commit/d66a07e202e7f1c611a90da6d85ada7f4ca9e357 resolves this issue.

**Comment by praveen-pk** [comment]
I manually TAP devices (with multiqueue support) and attached them to libvirt created virbr0 bridge. I started 3 VMs with cloud-hypervisor and all 3 of them get IP addresses from dnsmasq just fine.

I will check if TAP devices are being incorrectly managed in libvirt, in multi-vm scenario.

**Comment by russell-islam** [comment]
> For clarification: Are you expecting the virtio-net interfaces for each VM to be connected to the same bridge with a dhcp server running against that bridge with all the VMs getting dhcp leases from the same subnet?

Yes. Libvirt creates the tap interface for each VM and adds the interface to the libvirt's default bridge(virbr0). The problem we are facing that inside the guest the network interface is not getting the DHCP IP. Tap interface is sending the response but the guest is not able to process it.

**Comment by praveen-pk** [comment]
In current implementation, which needs to be fixed, both libvirt and clh create and manage queues/FDs. This is causing discrepancy between queue pairs of TAP device on host to that in the guest. This discrepancy could be causing the packet drops reported by in this bug.

Following is what I am seeing on guest and host:

```
# On Host
praween@praween-ubuntu-KVM:~$  ls /sys/devices/virtual/net/vnet0/queues/
rx-0  rx-1  rx-2  rx-3  rx-4  rx-5  tx-0  tx-1  tx-2  tx-3  tx-4  tx-5
# In Guest
testuser@chtestvm:~$ ls /sys/devices/pci0000\:00/0000\:00\:04.0/virtio3/net/ens4/queues/
rx-0  rx-1  tx-0  tx-1
# Networking configuration of guest
<interface type='network'>
        <mac address='00:11:22:33:46:11'/>
        <driver queues='4' iommu='on'/>
      <source network='default' bridge='virbr0'/>
      <model type='virtio'/>
    </interface>

```

If libvirt solely creates queue pairs and passes FDs to cloud-hypervisor, guest reboots cause the TAP interface to be deleted: https://github.com/cloud-hypervisor/libvirt/issues/65. That is why https://github.com/cloud-hypervisor/libvirt/commit/3a246c5e91fc8defa35ee72ffd69b33d1e9da5cf patch was added later. Unfortunately, this patch was not fully tested, which resulted in this bug.

Because of how ioctl for TAP is defined, libvirt cannot create the TAP device and let cloud-hypervisor enable queue pairs on the interface. 

Considering the above details, following are the possible approaches to fix this issue:

1) cloud-hypervisor will take control over creating & managing TAP devices. Whenever a new TAP device is created, clh notifies libvirt via monitor interface. libvirt will hook up the TAP device to the configured bridge. This approach would through guest reboots.

2) cloud-hypervisor would preserve TAP FDs across guest reboots. This will allow us to fallback to the original implementation of passing FDs to cloud-hypervisor

**Comment by russell-islam** [comment]
If we follow approach 1 we will be modifying the code flow that QEMU does today.

**Comment by russell-islam** [comment]
@rbradford Any suggestions?

**Comment by rbradford** [comment]
I would go with approach 1. Please create a PR with an event reported at the point that you want.

**Comment by sboeuf** [comment]
@praveen-pk 

Regarding the following
```
# On Host
praween@praween-ubuntu-KVM:~$  ls /sys/devices/virtual/net/vnet0/queues/
rx-0  rx-1  rx-2  rx-3  rx-4  rx-5  tx-0  tx-1  tx-2  tx-3  tx-4  tx-5
# In Guest
testuser@chtestvm:~$ ls /sys/devices/pci0000\:00/0000\:00\:04.0/virtio3/net/ens4/queues/
rx-0  rx-1  tx-0  tx-1
# Networking configuration of guest
<interface type='network'>
        <mac address='00:11:22:33:46:11'/>
        <driver queues='4' iommu='on'/>
      <source network='default' bridge='virbr0'/>
      <model type='virtio'/>
    </interface>
```
Could you try to create enough vCPUs (8 in this particular case) so that the guest will create enough queues? I agree if the host tries to send some packets using one of the TAP queue but the guest didn't activate all of them, we might run into some issues.

@rbradford I'm wondering why all the threads are "activated" on the host side even if the guest is only using a subset of the queues?

**Comment by rbradford** [comment]
> @rbradford I'm wondering why all the threads are "activated" on the host side even if the guest is only using a subset of the queues?

Do you mean you see the logs from the VMM saying that it is activated? Because if you do that means the guest has activated them.

**Comment by sboeuf** [comment]
@praveen-pk the difference in number of queues that you're seeing between guest and host is because in the guest you're looking at queues in terms of Linux networking. They are not VIRTIO queues as you could expect.
So that's why the virtio queues are all enabled and functioning, and the guest uses them all, but this is not what you can get from `ls /sys/devices/pci0000\:00/0000\:00\:04.0/virtio3/net/ens4/queues`.

**Comment by rbradford** [comment]
> I would go with approach 1. Please create a PR with an event reported at the point that you want.

You can also pursue option 2 if you like. Cloud Hypervisor now contains this fix: https://github.com/cloud-hypervisor/cloud-hypervisor/pull/2516

**Comment by sboeuf** [comment]
@russell-islam @praveen-pk I think option 2 would be the preferred approach, and as mentioned by @rbradford it should now support guest reboots.

**Comment by rbradford** [comment]
I think the source of your issues is that this change: https://github.com/cloud-hypervisor/libvirt/commit/3a246c5e91fc8defa35ee72ffd69b33d1e9da5cf from passing fds to tap interface name didn't also stop opening the tap interface on the libvirt side i.e. the same tap was opened by CH and by libvirt. If names are to be used the tap shouldn't be opened by libvirt.

**Comment by praveen-pk** [comment]
> I think the source of your issues is that this change: [3a246c5](https://github.com/cloud-hypervisor/libvirt/commit/3a246c5e91fc8defa35ee72ffd69b33d1e9da5cf) from passing fds to tap interface name didn't also stop opening the tap interface on the libvirt side i.e. the same tap was opened by CH and by libvirt. If names are to be used the tap shouldn't be opened by libvirt.

+1 Robert. That is the source of the issue. That is why we wanted to put the ownership solely on cloud-hypervisor or libvirt for creating and managing queue pairs. 

Now that https://github.com/cloud-hypervisor/cloud-hypervisor/pull/2516 is committed, we will revert back to libvirt creating the  queuepairs and passing the FDs to cloud-hypervisor.

**Comment by bryteise** [comment]
Fixed in #86
